### PR TITLE
Added signals handling by calling orphansKill for signals: SIGTERM, SIGHUP, SIGPIPE

### PR DESCRIPTION
**Pull request by Hojang2**

### CODE REVIEW

**Comment by Hojang2** [comment]
Because after killing the main process, processes like dnf were still running. This PR fixes this problem.

**Comment by Conan-Kudo** [comment]
This shouldn’t be the case if you’re using nspawn…

**Comment by xsuchy** [comment]
This is trying to address https://github.com/rpm-software-management/mock/issues/153

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,4 @@ bugzillas
 test-suite.log
 tests/runtests.sh.log
 tests/runtests.sh.trs
+.idea/
```
Comment (xsuchy) [inline .gitignore]: Can you put this in separate commit with the reasoning why you are adding it?

File: mock/py/mock.py [hunk 2]
```diff
@@ -666,6 +674,8 @@ def main():
     util.setup_host_resolv(config_opts)
 
     buildroot = Buildroot(config_opts, uidManager, state, plugins, bootstrap_buildroot)
+    signal.signal(signal.SIGTERM, partial(handle_sigterm, buildroot))
+    print("************************* Signal handler active *********************")
```
Comment (xsuchy) [inline mock/py/mock.py]: log.info() would be much better here. Without those asterisks.

File: mock/py/mock.py [hunk 3]
```diff
@@ -397,6 +399,12 @@ def command_parse():
     return (options, args)
 
 
+def handle_sigterm(buildroot, number, frame):
+    print("\nReceived SIGTERM activating orphansKill")
```
Comment (xsuchy) [inline mock/py/mock.py]: log.info() instead of print.

File: mock/py/mock.py [hunk 4]
```diff
@@ -948,6 +958,7 @@ def run_command(options, args, config_opts, commands, buildroot, state):
     except (KeyboardInterrupt,):
         exitStatus = 7
         log.error("Exiting on user interrupt, <CTRL>-C")
+        #  TODO: Make this work util.orphansKill(buildroot.make_chroot_path())
```
Comment (xsuchy) [inline mock/py/mock.py]: can you just call it? :)

File: mock/py/mock.py [hunk 5]
```diff
@@ -666,6 +674,8 @@ def main():
     util.setup_host_resolv(config_opts)
 
     buildroot = Buildroot(config_opts, uidManager, state, plugins, bootstrap_buildroot)
+    signal.signal(signal.SIGTERM, partial(handle_sigterm, buildroot))
```
Comment (xsuchy) [inline mock/py/mock.py]: Hmm, should we catch/process SIGABRT as well?

Comment (kdudka) [reply mock/py/mock.py]: On which occasions (other than internal error) does mock get `SIGABRT`?

I think it would make sense handle for example `SIGINT`, which is usually received when mock runs in the foreground process group and the user presses `^C`.

Comment (praiskup) [reply mock/py/mock.py]: E.g. `copr-rpmbuild` sends (on timeout) SIGINT down to mock as well.

Comment (Hojang2) [reply mock/py/mock.py]: Well, SIGINT is interpreted as KeyInterruptError in python and is already caught. As I mentioned below after KeyInterruptError all processes are killed, but it is not my work. If someone would test this too, it will be better.

Comment (kdudka) [reply mock/py/mock.py]: Indeed, `SIGINT` seems to be handled in recent versions of mock.

What about `SIGHUP` and `SIGPIPE`?

Try the following:
1. run `mock --chroot 'echo sleeping... && sleep 1024' &`
2. wait till `sleeping...` appears in your terminal
3. run `kill -SIGHUP $!` or `kill -SIGPIPE $!`
4. run `ps -A | grep sleep`

Comment (xsuchy) [reply mock/py/mock.py]: SIGHUP is a good idea.
SIGPIPE - I am not sure I want to encourage any use of pipes with mock. It has always been recommended to use logs with mock. Ẅith stdout, mock's code assumes human on the other side.

Comment (praiskup) [reply mock/py/mock.py]: But copr, koji and covscan are working with live logs...  which actually has potential to get broken pipes.

Comment (praiskup) [reply mock/py/mock.py]: That said - speaking about SIGPIPE - it would be nice to have at least some was to tell the mock that on stdout/stderr isn't consumed by human.

**Comment by kdudka** [comment]
> SIGPIPE - I am not sure I want to encourage any use of pipes with mock. It has always been recommended to use logs with mock.

csmock has been using pipes with mock on both stdout and stderr since forever.

> With stdout, mock's code assumes human on the other side.

This could have been a valid expectation in late '90s.  In 2019 everything needs to be automated!

**Comment by xsuchy** [comment]
At least since MSimacek Google Summer of Code work, there is a strong push in code to use separate logs. In Copr we use live log, but they are consumed by humans. But yeah, if csmock uses pipes, we can catch SIGPIPE as well.

**Comment by kdudka** [comment]
OK, thank you for considering it!

Slightly off-topic but, at least `mock --shell` should never print anything on stdout because it could interfere with output of the command wrapped by mock.  For example, csmock uses the following command to get data out of mock's chroot:
```
mock --shell 'tar -c /builddir/cswrap-capture.err /builddir/cov ...' | tar -x ...
```

File: mock/py/mock.py [hunk 6]
```diff
@@ -397,6 +399,12 @@ def command_parse():
     return (options, args)
 
 
+def handle_signals(buildroot, number, frame):
+    log.info("\nReceived SIGTERM activating orphansKill")
+    util.orphansKill(buildroot.make_chroot_path())
+    sys.exit(7)
```
Comment (praiskup) [inline mock/py/mock.py]: Exit status 7 isn't documented:
https://github.com/rpm-software-management/mock/blob/master/mock/py/mockbuild/exception.py#L26

Comment (kdudka) [reply mock/py/mock.py]: Good point.  The exit status should be `128 + signum` (to be compatible with shell interpreters).

Comment (Hojang2) [reply mock/py/mock.py]: So, if the signum is 13, then exit status should be 141?

Comment (praiskup) [reply mock/py/mock.py]: Since those exit status aren't unused, it sounds good to me.  Btw., is it safe to actually call exit in signal handler?

Comment (kdudka) [reply mock/py/mock.py]: In C run-time `_exit()` and `_Exit()` are async-signal-safe functions (because they do not call `atexit()`-registered callbacks).  No idea about python though...

Comment (Hojang2) [reply mock/py/mock.py]: > Exit status 7 isn't documented:
> https://github.com/rpm-software-management/mock/blob/master/mock/py/mockbuild/exception.py#L26

should I document all possible exit codes with signals? like 141, 143, 129

Comment (Hojang2) [reply mock/py/mock.py]: in the exceptions file?

Comment (kdudka) [reply mock/py/mock.py]: I would prefer to have them documented symbolically.  For example, this is how `bash(1)` man page documents the codes:

> The return value of a simple command is its exit status, or 128+n if the command is terminated by signal n.

**Comment by xsuchy** [comment]
Yes, please document it in exception.py. I do not care if a symbolic way (128+X) or enumerate.

File: mock/py/mock.py [hunk 7]
```diff
@@ -397,6 +399,20 @@ def command_parse():
...
 
 
+def handle_signals(buildroot, number, frame):
+    signal = ""
+    if number == 1:
+        signal = "SIGHUP"
+    elif number == 13:
+        signal = "SIGPIPE"
+    elif number == 15:
+        signal = "SIGTERM"
```
Comment (xsuchy) [inline mock/py/mock.py]: This can be better done as
```
signal_names = {1: "SIGHUP", 
               13, "SIGPIPE",
...
}
#all this can be defined at the top of file. and then
log.info("\nReceived signal {} activating orphansKill".format(signal_name[number]))
```
btw original name of variable `signal` override the name of the module. Not a good idea.

**Comment by xsuchy** [comment]
LGTM, thank you
